res/controllers/midi-components-0.0.js

```diff
@@ -301,8 +309,48 @@
         shift: function () {
             this.inKey = 'hotcue_' + this.number + '_clear';
         },
+        getColor: function() {
+            if (this.colorIdKey !== undefined) {
+                return color.predefinedColorFromId(engine.getValue(this.group,this.colorIdKey));
+            } else {
+                return null;
+            }
+        },
+        output: function(value) {
+            var outval = this.outValueScale(value);
+            if (this.colorIdKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorIdKey));
+            } else {
+                this.send(outval);
+            }
+        },
+        outputColor: function (id) {
+            var color = this.colors[id];
```

Comment:
Why is `colors` a property of the Component? Why not simply access `colors.predefinedColorsList` here? Is there a use case for having a HotcueButton Component that has a color list different from `colors.predefinedColorsList`?

Reply:
Yes, It allows to establish a mapping between color ID and midi value to set an LED to that color. These can differ wildy between different hardware controller models.
Here's a real-world controller mapping that uses that property: https://github.com/mixxxdj/mixxx/pull/2111/commits/33889a50c63fdfd542d3b1851aaff9a3f49652ce

Reply:
exactly. Every controller responds with different colors to certain midi values. The `colors` array maps between the values the controller expects and the colorIDs in Mixxx. If the controller supports setting exact rgb values via sysex, no color mapping has to be defined (so it just copies `colors.predefinedColorsList`. However, the `sendRBG` method has to be implemented in that case which handles the construction and sending of the sysex message (since these also vary from controller to controller).

Reply:
Okay, that makes sense. So you would set `components.HotcueButton.colors.prototype` for the specific values for the hardware? This should be explained in the documentation. Also the color corresponding to each colorID index should be documented; otherwise, script writers would need to read the C++ source code to find them.

Reply:
> So you would set components.HotcueButton.colors.prototype for the specific values for the hardware?

Yes, either in the buttons (as shown in the sample code) or just the prototype directly.

> Also the color corresponding to each colorID index should be documented.

Of course, I thought they were already documented somewhere during the development of the PR that introduced the color set.

Reply:
> I thought they were already documented somewhere during the development of the PR that introduced the color set.

I'm not aware of such documentation.